fuel-p2p/src/orchestrator.rs

```diff
@@ -110,26 +109,157 @@ impl NetworkOrchestrator {
+}
+
+impl Service {
+    pub fn new(
+        p2p_config: P2PConfig,
+        db: Arc<dyn P2pDb>,
+        tx_request_event: Sender<P2pRequestEvent>,
+        rx_request_event: Receiver<P2pRequestEvent>,
+        tx_consensus: Sender<ConsensusBroadcast>,
+        tx_transaction: Sender<TransactionBroadcast>,
+        tx_block: Sender<BlockBroadcast>,
+    ) -> Self {
+        let network_orchestrator = NetworkOrchestrator::new(
+            p2p_config,
+            rx_request_event,
+            tx_consensus,
+            tx_transaction,
+            tx_block,
+            db,
+        );
+
+        Self {
+            join: Mutex::new(None),
+            network_orchestrator: Arc::new(Mutex::new(Some(network_orchestrator))),
+            tx_request_event,
+        }
+    }
+
+    pub async fn start(&self) -> anyhow::Result<()> {
+        let mut join = self.join.lock().await;
+
+        if join.is_none() {
+            if let Some(network_orchestrator) = self.network_orchestrator.lock().await.take() {
+                *join = Some(tokio::spawn(async { network_orchestrator.run().await }));
+
+                Ok(())
+            } else {
+                Err(anyhow!("Starting Network Orchestrator that is stopping"))
+            }
+        } else {
+            Err(anyhow!("Network Orchestrator already started"))
+        }
+    }
+
+    pub async fn stop(&self) -> Option<JoinHandle<()>> {
+        let join_handle = self.join.lock().await.take();
+
+        if let Some(join_handle) = join_handle {
+            let network_orchestrator = self.network_orchestrator.clone();
+            let _ = self.tx_request_event.send(P2pRequestEvent::Stop).await;
+            Some(tokio::spawn(async move {
+                if let Ok(res) = join_handle.await {
+                    *network_orchestrator.lock().await = res.ok();
+                }
+            }))
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use async_trait::async_trait;
+    use fuel_core_interfaces::model::{
+        BlockHeight, FuelBlock, FuelBlockConsensus, SealedFuelBlock,
+    };
+    use tokio::time::{sleep, Duration};
+
+    #[derive(Clone, Debug)]
+    struct FakeDb;
+
+    #[async_trait]
+    impl P2pDb for FakeDb {
+        async fn get_sealed_block(&self, _height: BlockHeight) -> Option<Arc<SealedFuelBlock>> {
+            let block = FuelBlock {
+                header: Default::default(),
+                transactions: vec![],
+            };
+
+            Some(Arc::new(SealedFuelBlock {
+                block,
+                consensus: FuelBlockConsensus {
+                    required_stake: 100_000,
+                    validators: Default::default(),
+                },
+            }))
+        }
+    }
+
+    #[tokio::test]
+    async fn start_stop_works() {
+        let mut p2p_config = P2PConfig::default_with_network("start_stop_works");
+        p2p_config.tcp_port = 4018; // an unused port
```

Comment:
nit: Would specifying port 0 to ask the OS for an available port be useful?

Reply:
Using something like Portpicker would be beneficial here as when running multithreaded tests unless each port is individually assigned there can be collision

Reply:
ah this is a good point!, I will try and add this to all of my p2p tests